[WIP] Port library to JAX.

chahak13 opened the pull request:
This PR aims to port the mpm functions to JAX so that the library can be used in differentiable tasks.

## TODO:
- [x] Nodes
- [x] Particles
- [x] Elements
- [x] Shape Functions
- [x] Meshes
- [x] Solvers

chahak13 committed:
Simple node class using jax to start porting.

chahak13 committed:
Add shape function notes

chahak13 committed:
Add tests for shape function

chahak13 committed:
Add files to be able to install diffmpm

Needed for running `pytest`.

kks32 commented on diffmpm/node.py:
```diff
@@ -1,23 +1,53 @@
+        return
 
+    def _reset_values(self):
+        self.position.fill(0)
```
I don't think we should reset the position.

chahak13 replied:
Yeah, that shouldn't be there.

kks32 commented on diffmpm/node.py:
```diff
@@ -1,23 +1,53 @@
     """
+    Nodes container class.
+
+    Keeps track of all values required for nodal points.
```
Should we define something like dimensions to keep track of?

chahak13 replied:
For higher dimensions? I thought about it. There are 2 ways we could go about it, either make n-D arrays or have 1D arrays with explicit indexing. Honestly, not very sure which will be easier so I just deferred this decision to later.

kks32 commented on diffmpm/shapefn.py:
```diff
@@ -1,16 +1,108 @@
-    """
-    Linear 1D shape function in natural coordinates
+
+class ShapeFn:
```
There is also GIMP shape functions even for 1D...but let's go ahead with this. We can think about GIMP shape functions later

chahak13 replied:
Ah okay. That should be easy to add via a kwarg so shouldn't be very difficult since gradient calculation is automatic so we'll just have to specify the function and it's transform.

kks32 commented on diffmpm/shapefn.py:
```diff
@@ -1,16 +1,108 @@
+        self.dim = dim
+        return
+
+    def shapefn(self, x):
```
Could you please use `xi` instead of `x`, since that's the variable we use to define natural coordinates.

chahak13 replied:
Okay, got it.

chahak13 committed:
Don't reset nodal position and change shapefn var

chahak13 committed:
Add first version of Material, Mesh, Particles

kks32 commented on diffmpm/mesh.py:
```diff
@@ -1,43 +1,111 @@
-from diffmpm import node
-from diffmpm import particle
 
 class Mesh1D:
```
Why this is Mesh1D while other classes don't have the dimension in their name?

chahak13 replied:
My try is to make the container classes such that for all dimensions, you can still call the same constructor, maybe with dimension as an argument. But at some point, we would want to handle things differently for different dimensions. Mesh seemed like a good place where this differentiation can be implemented. If we then want to have a consistent API for the user for Mesh like the other classes, we can create a parent class and make all these dimension specific classes as private classes. This was one of the things that I want to discuss today.

kks32 commented on diffmpm/mesh.py:
```diff
@@ -1,43 +1,111 @@
+        self.domain_size = domain_size
+        self.nelements = nelements
+        self.element_length = domain_size / nelements
+        self.elements = jnp.arange(nelements)
+        self.nodes = Nodes(nelements + 1)
```
Note: This assumes linear shape functions, not GIMP. We can worry about this later, just letting you know.

chahak13 replied:
Yeah, I'll change it to a better flow/initialization for different shape functions.

chahak13 committed:
Add length property to Nodes and Particles

chahak13 committed:
Add method to find element ids for particles

This also adds a file containing some basic tests for the Mesh class.

chahak13 committed:
Add function to update particles' natural coords

chahak13 committed:
Add docstrings

chahak13 committed:
Add boundary nodes info and change variable name

chahak13 committed:
Correct shapefn dims and add vector tests

kks32 commented on diffmpm/particle.py:
```diff
@@ -1,15 +1,66 @@
         self.mass = mass
-        self.x  = x 
-        self.material = material 
-        self.density = material.density 
```
Why did you remove density, we need to compute mass of particle

chahak13 replied:
Ignore previous comment. I made a mistake in reading the notation in one step. Adding density again.

chahak13 committed:
Reset node values using .at; particle mass array

chahak13 committed:
Add new update functions and tests

chahak13 committed:
Add function to get total force on nodes.

chahak13 committed:
Add density to particles

chahak13 committed:
Add few more update functions and tests

chahak13 committed:
Add particle -> node update functions and tests

chahak13 committed:
Use raw docstrings and change protected function

chahak13 committed:
Add first version of mesh.solve()

This commit also corrects a change in updates where
the shape function was to be calculated for natural coords (xi) insteaf
of physical coords (x). Furthermore, internal force update needed
gradient of shape function and not shape function itself.

chahak13 committed:
Many changes to make solver JIT compatible.

This needed various changes in the constructors of the different classes
along with registering them as Pytree nodes in the jax registry so that
I can pass around `self` as an argument. The JIT solver still needs work
for branching and conditions but just for USF it works.

chahak13 committed:
Add control flow for USF, USL, MUSL
